xbmc/utils/StringUtils.cpp

```diff
@@ -1123,6 +1123,171 @@ int64_t StringUtils::AlphaNumericCompare(const wchar_t* left, const wchar_t* rig
+  {
+    // Check if we have numerical values, compare only up to 15 digits
+    if (isdigit(zA[i]) && isdigit(zB[j]))
+    {
+      lnum = 0;
+      k = 0;
+      while (i < nKey1 && isdigit(zA[i]) && k < 15)
+      {
+        lnum *= 10;
+        lnum += zA[i] - '0';
+        k++;
+        i++;
+      }
+      rnum = 0;
+      k = 0;
+      while (j < nKey2 && isdigit(zB[j]) && k < 15)
+      {
+        rnum *= 10;
+        rnum += zB[j] - '0';
+        k++;
+        j++;
+      }
+      // do we have numbers?
+      if (lnum != rnum)
+      { // yes - and they're different!
+        return lnum - rnum;
+      }
+      continue; // i and j already advanced through digits
+    }
+    // Put ascii punctuation and symbols e.g. !#$&()*+,-./:;<=>?@[\]^_ `{|}~ before the other
+    // alphanumeric ascii, rather than some being mixed between the numbers and letters, and
+    // above all other unicode letters, symbols and punctuation.
+    // (Locale collation of these chars varies across platforms)
+    lsym = (zA[i] < 128 && zA[i] > 'z') || (zA[i] < 'a' && zA[i] > 'Z') ||
+           (zA[i] < 'A' && zA[i] > '9') || (zA[i] < '0' && zA[i] >= 32);
+    rsym = (zB[j] < 128 && zB[j] > 'z') || (zB[j] < 'a' && zB[j] > 'Z') ||
+           (zB[j] < 'A' && zB[j] > '9') || (zB[j] < '0' && zB[j] >= 32);
+    if (lsym && !rsym)
+      return -1;
+    if (!lsym && rsym)
+      return 1;
+    if (lsym && rsym)
+    {
+      if (zA[i] != zB[j])
+        return zA[i] - zB[j];
+      else
+      {
+        i++;
+        j++;
+        continue;
+      }
+    }
+    //Decode single (1 to 4 bytes) UTF8 character to Unicode
+    lc = Utf8ToUnicode(&zA[i], nKey1 - i, bytes);
+    i += bytes;
+    rc = Utf8ToUnicode(&zB[j], nKey2 - j, bytes);
+    j += bytes;
+    if (!g_langInfo.UseLocaleCollation())
+    {
+      // Apply case sensitive accent folding collation to non-ascii chars.
+      // This mimics utf8_general_ci collation, and provides simple collation of LATIN-1 chars
+      // for any platform that doesn't have a language specific collate facet implemented
+      if (lc > 128)
+        lc = GetCollationWeight(lc);
+      if (rc > 128)
+        rc = GetCollationWeight(rc);
+    }
+    // Caseless comparison so convert ascii upper case to lower case
+    if (lc >= 'A' && lc <= 'Z')
+      lc += 'a' - 'A';
+    if (rc >= 'A' && rc <= 'Z')
+      rc += 'a' - 'A';
+
+    if (lc != rc)
+    {
+      if (!g_langInfo.UseLocaleCollation())
+        // Compare unicode (having applied accent folding collation to non-ascii chars).
+        return lc - rc;
+      else
+      {
+        // Fetch collation facet from locale to do comparison of wide char although on some
+        // platforms this is not langauge specific but just compares unicode
+        const std::collate<wchar_t>& coll =
+            std::use_facet<std::collate<wchar_t>>(g_langInfo.GetSystemLocale());
+        int cmp_res = 0;
+        cmp_res = coll.compare(&lc, &lc + 1, &rc, &rc + 1);
```

Comment:
Can this be combined with the next line to reduce redundancy?
```suggestion
        int cmp_res = coll.compare(&lc, &lc + 1, &rc, &rc + 1);
```